src/Common/src/System/Security/Cryptography/Pkcs12Kdf.cs

```diff
@@ -0,0 +1,280 @@
+    {
+        private const byte CipherKeyId = 1;
+        private const byte IvId = 2;
+        private const byte MacKeyId = 3;
+
+        // This is a dictionary representation of the table in
+        // https://tools.ietf.org/html/rfc7292#appendix-B.2
+        private static readonly Dictionary<HashAlgorithmName, Tuple<int, int>> s_uvLookup =
+            new Dictionary<HashAlgorithmName, Tuple<int, int>>
+            {
+                { HashAlgorithmName.MD5, Tuple.Create(128, 512) },
+                { HashAlgorithmName.SHA1, Tuple.Create(160, 512) },
+                { HashAlgorithmName.SHA256, Tuple.Create(256, 512) },
+                { HashAlgorithmName.SHA384, Tuple.Create(384, 1024) },
+                { HashAlgorithmName.SHA512, Tuple.Create(512, 1024) },
+            };
+
+        internal static void DeriveCipherKey(
+            ReadOnlySpan<char> password,
+            HashAlgorithmName hashAlgorithm,
+            int iterationCount,
+            ReadOnlySpan<byte> salt,
+            Span<byte> destination)
+        {
+            Derive(
+                password,
+                hashAlgorithm,
+                iterationCount,
+                CipherKeyId,
+                salt,
+                destination);
+        }
+
+        internal static void DeriveIV(
+            ReadOnlySpan<char> password,
+            HashAlgorithmName hashAlgorithm,
+            int iterationCount,
+            ReadOnlySpan<byte> salt,
+            Span<byte> destination)
+        {
+            Derive(
+                password,
+                hashAlgorithm,
+                iterationCount,
+                IvId,
+                salt,
+                destination);
+        }
+
+        internal static void DeriveMacKey(
+            ReadOnlySpan<char> password,
+            HashAlgorithmName hashAlgorithm,
+            int iterationCount,
+            ReadOnlySpan<byte> salt,
+            Span<byte> destination)
+        {
+            Derive(
+                password,
+                hashAlgorithm,
+                iterationCount,
+                MacKeyId,
+                salt,
+                destination);
+        }
+
+        private static void Derive(
+            ReadOnlySpan<char> password,
+            HashAlgorithmName hashAlgorithm,
+            int iterationCount,
+            byte id,
+            ReadOnlySpan<byte> salt,
+            Span<byte> destination)
+        {
+            // https://tools.ietf.org/html/rfc7292#appendix-B.2
+            Debug.Assert(iterationCount >= 1);
+
+            if (!s_uvLookup.TryGetValue(hashAlgorithm, out Tuple<int, int> uv))
+            {
+                throw new CryptographicException(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithm.Name);
+            }
+
+            (int u, int v) = uv;
+
+            //  1. Construct a string, D (the "diversifier"), by concatenating v/8 copies of ID.
+            int vBytes = v >> 3;
+            Span<byte> D = stackalloc byte[vBytes];
```

Comment:
should we check for max size here?

Reply:
The biggest defined `v` is 1024 (making `vBytes` 128).  But I can add an assert, in case the RFC backpedals and adds more options (despite calling this KDF 'legacy').